### PR TITLE
Hosted services

### PR DESCRIPTION
**Pull request by scareything**
Hosted services working for tcp and udp server configs.

### CODE REVIEW

File: lib/ziti_tunneler.c [hunk 1]
```diff
@@ -136,13 +143,13 @@ void ziti_tunneler_stop_intercepting(tunneler_context tnlr_ctx, const char *serv
 }
 
 /** called by tunneler application when data is read from a ziti connection */
-int ziti_tunneler_write(tunneler_io_context *tnlr_io_ctx, const void *data, size_t len) {
+ssize_t ziti_tunneler_write(tunneler_io_context *tnlr_io_ctx, const void *data, size_t len) {
```
Comment (dovholuknf) [inline lib/ziti_tunneler.c:146]: I think using ssize_t isn't particularly windows friendly. @ekoby, correct?

File: lib/ziti_tunneler.c [hunk 2]
```diff
@@ -228,8 +235,9 @@ static struct raw_pcb * init_protocol_handler(u8_t proto, raw_recv_fn recv_fn, v
 }
 
 static void run_packet_loop(uv_loop_t *loop, tunneler_context tnlr_ctx) {
-    if (tnlr_ctx->opts.ziti_close == NULL || tnlr_ctx->opts.ziti_dial == NULL || tnlr_ctx->opts.ziti_write == NULL) {
-        ZITI_LOG(ERROR, "ziti_* callback options cannot be null");
+    if (tnlr_ctx->opts.ziti_close == NULL || tnlr_ctx->opts.ziti_dial == NULL ||
+        tnlr_ctx->opts.ziti_write == NULL || tnlr_ctx->opts.ziti_host_v1 == NULL) {
+        ZITI_LOG(ERROR, "ziti_sdk_* callback options cannot be null");
```
Comment (dovholuknf) [inline lib/ziti_tunneler.c:240]: Small nit but it'd be friendly if this logged exactly which are missing. Not a big deal

File: lib/ziti_tunneler_cbs.c [hunk 3]
```diff
@@ -82,4 +92,264 @@ static void on_ziti_write(ziti_connection ziti_conn, ssize_t len, void *ctx) {
...
+    if (io_ctx == NULL) {
+        ZITI_LOG(ERROR, "null io_ctx");
+        //free(buf->base); TODO free this here?
+        return;
+    }
+
+    if (nread > 0) {
+        ziti_write(io_ctx->client, buf->base, nread, on_hosted_ziti_write, buf->base);
+    } else {
+        ZITI_LOG(INFO, "stuff");
```
Comment (dovholuknf) [inline lib/ziti_tunneler_cbs.c:196]: Stuff indeed... :)

File: lib/ziti_tunneler_cbs.c [hunk 4]
```diff
@@ -82,4 +92,264 @@ static void on_ziti_write(ziti_connection ziti_conn, ssize_t len, void *ctx) {
...
+        ziti_close(&io_ctx->client);
+    }
+}
+
+/** called by libuv when a hosted UDP server sends data to a client */
+static void on_hosted_udp_server_data(uv_udp_t* handle, ssize_t nread, const uv_buf_t* buf, const struct sockaddr* addr, unsigned flags) {
+    struct hosted_io_ctx_s *io_ctx = handle->data;
+    if (nread > 0) {
+        ziti_write(io_ctx->client, buf->base, nread, on_hosted_ziti_write, buf->base);
+    }
```
Comment (dovholuknf) [inline lib/ziti_tunneler_cbs.c:206]: Don't care if nread is <= 0?

File: lib/ziti_tunneler_cbs.c [hunk 5]
```diff
@@ -82,4 +92,264 @@ static void on_ziti_write(ziti_connection ziti_conn, ssize_t len, void *ctx) {
...
+    struct hosted_io_ctx_s *io_ctx = c->handle->data;
+    ZITI_LOG(INFO, "connected to server for client %p: %p", c->handle->data, c);
+    uv_read_start((uv_stream_t *) io_ctx->server.tcp, alloc_buffer, on_hosted_tcp_server_data);
+}
+
+/** called by ziti sdk when a client connection is established (or fails) */
+static void on_hosted_client_connect_complete(ziti_connection clt, int status) {
+    ZITI_LOG(INFO, "client %p connected to hosted service", clt);
+    if (status == ZITI_OK) {
```
Comment (dovholuknf) [inline lib/ziti_tunneler_cbs.c:228]: No else clause for when this is not ok?

File: lib/ziti_tunneler_cbs.c [hunk 6]
```diff
@@ -82,4 +92,264 @@ static void on_ziti_write(ziti_connection ziti_conn, ssize_t len, void *ctx) {
...
+        struct hosted_service_ctx_s *service_ctx = ziti_conn_data(clt);
+        if (service_ctx == NULL) {
+            ZITI_LOG(DEBUG, "null service_ctx");
+            ziti_close(&clt);
```
Comment (dovholuknf) [inline lib/ziti_tunneler_cbs.c:232]: Should this return here?

File: lib/ziti_tunneler_cbs.c [hunk 7]
```diff
@@ -82,4 +92,264 @@ static void on_ziti_write(ziti_connection ziti_conn, ssize_t len, void *ctx) {
...
+
+    if (status != ZITI_OK) {
+        ZITI_LOG(ERROR, "unable to host service %s: %s", host_ctx->service_name, ziti_errorstr(status));
+        free_hosted_service_ctx(host_ctx);
+    }
+}
+
+/** called by the tunneler sdk when a hosted service becomes available */
+void ziti_sdk_c_host_v1(ziti_context ziti_ctx, uv_loop_t *loop, const char *service_name, const char *proto, const char *hostname, int port) {
+    if (service_name == NULL) {
```
Comment (dovholuknf) [inline lib/ziti_tunneler_cbs.c:318]: It'd be friendly imo to log all these and exit if any triggered.

File: programs/ziti-tunnel-app/ziti-tunnel-app.c [hunk 8]
```diff
@@ -28,18 +28,34 @@ static ziti_options OPTS = {
...
+            } else {
+                ZITI_LOG(INFO, "service %s lacks ziti-tunneler-server.v1 config; not hosting", service->name);
+            }
         }
-        ZITI_LOG(DEBUG, "ziti_service_get_config rc: %d", rc);
     } else if (status == ZITI_SERVICE_UNAVAILABLE) {
         ZITI_LOG(INFO, "service unavailable: %s", service->name);
-        ziti_tunneler_stop_intercepting(tnlr_ctx, service->id);
+        ziti_tunneler_stop_intercepting(tnlr_ctx, service->name);
     }
```
Comment (dovholuknf) [inline programs/ziti-tunnel-app/ziti-tunnel-app.c:59]: What about a catch all else?

File: lib/ziti_tunneler.c (hunk 1 above)
Comment (ekoby) [reply lib/ziti_tunneler.c:146]: correct in general, but I both libuv and lwip declare proper type and we get it

**Pull request merged**